sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,502 @@
+
+    abstract Builder builder();
+
+    public void populateDisplayData(DisplayData.Builder builder) {
+      builder.addIfNotNull(DisplayData.item("hosts", getHosts().toString()));
+      builder.addIfNotNull(DisplayData.item("keyspace", getKeyspace()));
+      builder.add(DisplayData.item("port", getPort()));
+    }
+
+    Cluster getCluster() {
+      LOG.debug("Connecting to Cassandra cluster");
+      return Cluster.builder()
+          .addContactPoints(getHosts().toArray(new String[0]))
+          .withPort(getPort())
+          .build();
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setHosts(List<String> hosts);
+
+      abstract Builder setKeyspace(String keyspace);
+
+      abstract Builder setPort(int port);
+
+      abstract ConnectionConfiguration build();
+    }
+  }
+
+  /**
+   * A {@link PTransform} to read data from Apache Cassandra. See {@link CassandraIO} for more
+   * information on usage and configuration.
+   */
+  @AutoValue
+  public abstract static class Read<T> extends PTransform<PBegin, PCollection<T>> {
+
+    @Nullable abstract ConnectionConfiguration connectionConfiguration();
+    @Nullable abstract String table();
+    @Nullable abstract Class<T> entity();
+    @Nullable abstract Coder<T> coder();
+    @Nullable abstract CassandraService<T> cassandraService();
+    abstract Builder<T> builder();
+
+    public Read<T> withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(
+          connectionConfiguration != null,
+          "CassandraIO.read()"
+              + ".withConnectionConfiguration(configuration) called with null configuration");
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    public Read<T> withTable(String table) {
+      checkArgument(table != null, "CassandraIO.read().withTable(table) called with null table");
+      return builder().setTable(table).build();
+    }
+
+    public Read<T> withEntity(Class<T> entity) {
+      checkArgument(entity != null, "CassandraIO.read().withEntity(entity) called with null "
+          + "entity");
+      return builder().setEntity(entity).build();
+    }
+
+    public Read<T> withCoder(Coder<T> coder) {
+      checkArgument(coder != null, "CassandraIO.read().withCoder(coder) called with null coder");
+      return builder().setCoder(coder).build();
+    }
+
+    public Read<T> withCassandraService(CassandraService<T> cassandraService) {
+      checkArgument(cassandraService != null, "CassandraIO.read().withCassandraService(service)"
+          + " called with null service");
+      return builder().setCassandraService(cassandraService).build();
+    }
+
+    @Override
+    public PCollection<T> expand(PBegin input) {
+      return input.apply(org.apache.beam.sdk.io.Read.from(
+          new CassandraSource<T>(this, null,
+              new SerializableFunction<PipelineOptions, CassandraService<T>>() {
```

Comment:
Why do you need this SerializableFunction? Seems like CassandraSource already takes an instance of the CassandraIO.Read transform and could just call .getCassandraService() on it.